File: plugins/okf/hooks/curate.rb
# frozen_string_literal: true

# The plugin's PostToolUse curation hook: after Claude writes or edits a file
# inside an OKF bundle, run `okf validate` + `okf lint` on that bundle and hand
# the relevant findings back as additional context. Deterministic — the gem's
# own checks are the detector, no LLM in the loop.
#
# Contract (Claude Code hooks):
#   stdin   the event JSON (tool_input.file_path, session_id, …)
#   stdout  optional JSON: { "hookSpecificOutput": { "additionalContext": … } }
#   exit 0  always — curation must never turn an unrelated edit into an error.
#
# Plain Ruby, stdlib only, same 2.4 floor as the gem: the hook runs on whatever
# Ruby the machine already has, before the gem is even installed.
#
# Off switches, all config-free (OKF keeps no config file):
#   OKF_CURATE_DISABLED   truthy → the hook does nothing, on any edit;
#   OKF_CURATE_QUIET      truthy → drop the advisory "install the CLI" nudge,
#                         but still surface real findings;
#   <!-- okf-disable -->  in the edited file → silence curation for that file.
# An env var overrides everything; the file marker is the per-file escape hatch.

require "json"
require "fileutils"
require "open3"
require "tmpdir"

module OKF
  module PluginHook
    class Curate
      # Keep the feedback scannable; `/okf:gem curate` exists for the full report.
      MAX_LINES = 12

      # A file opts out with an `<!-- okf-disable -->` comment anywhere in it.
      # Only whole-file: lint findings are per-concept, not line-anchored, so a
      # line-level marker would have no line to point at.
      DISABLE_MARKER = /<!--[^>]*\bokf-disable\b[^>]*-->/.freeze

      def self.run(stdin: $stdin, stdout: $stdout)
        new.run(stdin, stdout)
      end

      # The CLI runner is injectable so tests drive the hook without a gem
      # install: it takes (verb, root) and returns the CLI's stdout.
      def initialize(runner: nil)
        @runner = runner || method(:run_cli)
      end

      def run(stdin, stdout)
        return 0 if disabled?

        event = parse(stdin) or return 0
        file = edited_markdown(event) or return 0
        return 0 if suppressed?(file)

        root = bundle_root(file) or return 0

        reports = fetch_reports(root)
        return missing_cli(event, stdout) if reports.nil?

        context = context_for(reports, relative(file, root))
        unless context.nil?
          emit(stdout, "hookSpecificOutput" => {
            "hookEventName" => "PostToolUse", "additionalContext" => context
          })
        end
        0
      end

      # ── the off switches ─────────────────────────────────────────────────────

      def disabled?
        env_flag?("OKF_CURATE_DISABLED")
      end

      def quiet?
        env_flag?("OKF_CURATE_QUIET")
      end

      # Set-and-truthy: any value other than the usual "off" spellings turns it on.
      def env_flag?(name)
        value = ENV[name].to_s.strip.downcase
        !value.empty? && !%w[0 false no off].include?(value)
      end

      # True when the edited file carries the opt-out marker. A read failure is
      # treated as "no marker" — the off switch must never become an error.
      def suppressed?(file)
        return false unless File.file?(file)

        DISABLE_MARKER.match?(File.read(file))
      rescue StandardError
        false
      end

      # ── the gate: cheap enough to run on every edit ──────────────────────────

      # The edited file when it is markdown, nil otherwise (other tools, other
      # file types, missing paths — all silently out of scope).
      def edited_markdown(event)
        input = event["tool_input"]
        path = input.is_a?(Hash) ? input["file_path"] : nil
        path.is_a?(String) && path.end_with?(".md") ? File.expand_path(path) : nil
      end

      # Walk up from the edited file looking for a bundle root: the directory
      # whose index.md opens a frontmatter block carrying okf_version — only the
      # bundle-root index has frontmatter at all (§9.3), so a nested index never
      # matches. A stat per level and one short read; no gem process.
      def bundle_root(file)
        dir = File.dirname(file)
        loop do
          return dir if root_index?(File.join(dir, "index.md"))

          parent = File.dirname(dir)
          return nil if parent == dir

          dir = parent
        end
      end

      def root_index?(path)
        return false unless File.file?(path)

        head = File.open(path, "rb") { |io| io.read(512) }.to_s
        head.start_with?("---") && head.include?("okf_version")
      end

      # ── the detectors: the okf CLI itself ────────────────────────────────────

      # Both reports parsed, or nil when the CLI is not installed. A CLI that
      # exists but misbehaves degrades to empty reports, never to an error.
      def fetch_reports(root)
        { validation: report("validate", root), lint: report("lint", root) }
      rescue Errno::ENOENT
        nil
      end

      def report(verb, root)
        JSON.parse(@runner.call(verb, root).to_s)
      rescue JSON::ParserError
        {}
      end

      def run_cli(verb, root)
        out, _err, _status = Open3.capture3("okf", verb, root, "--json")
        out
      end

      # ── findings → context ───────────────────────────────────────────────────

      # The curation text for this edit, or nil when there is nothing to say:
      # every conformance error (bundle-wide, always hard), plus the validate
      # warnings and lint findings that concern the edited file.
      def context_for(reports, rel)
        errors = Array(reports[:validation]["errors"])
        warnings = Array(reports[:validation]["warnings"]).select { |w| w["path"] == rel }
        findings = Array(reports[:lint]["findings"]).select { |f| concerns?(f, rel) }
        return nil if errors.empty? && warnings.empty? && findings.empty?

        build_context(errors, warnings, findings)
      end

      # A lint finding concerns the edit when it points at the edited file, or —
      # for backlog findings like missing_concept — when the edited concept is
      # among the sources demanding a target that does not exist.
      def concerns?(finding, rel)
        return true if finding["path"] == rel

        metric = finding["metric"]
        sources = metric.is_a?(Hash) ? Array(metric["sources"]) : []
        sources.include?(rel.sub(/\.md\z/, ""))
      end

      def build_context(errors, warnings, findings)
        lines = errors.map { |e| "  ✗ error #{e["path"]}: #{e["message"]}" } +
                warnings.map { |w| "  ! warn #{w["path"]}: #{w["message"]}" } +
                findings.map { |f| "  #{f["severity"] == "warn" ? "!" : "·"} lint/#{f["check"]} #{f["path"]}: #{f["message"]}" }

        shown = lines.take(MAX_LINES)
        shown << "  … #{lines.size - MAX_LINES} more (run /okf:gem curate for the full report)" if lines.size > MAX_LINES
        verdict = errors.empty? ? "Advisory curation debt — settle what this task touches." : "Fix the error(s) before finishing; the rest is curation debt."
        ([ "okf curation — #{summary(errors, warnings, findings)}:" ] + shown + [ verdict ]).join("\n")
      end

      def summary(errors, warnings, findings)
        parts = []
        parts << pluralize(errors.size, "error") unless errors.empty?
        parts << pluralize(warnings.size, "warning") unless warnings.empty?
        parts << pluralize(findings.size, "lint finding") unless findings.empty?
        parts.join(", ")
      end

      def pluralize(count, noun)
        count == 1 ? "#{count} #{noun}" : "#{count} #{noun}s"
      end

      # `okf` not installed: say so once per session (a marker file keyed by the
      # session id), then stay silent — never an error per edit. Quiet mode drops
      # the nudge entirely; the CLI is still absent, there is just nothing to fix
      # per edit.
      def missing_cli(event, stdout)
        return 0 if quiet?

        session = event["session_id"].to_s.gsub(/[^\w-]/, "_")
        data_dir = ENV["PLUGIN_DATA"].to_s
        data_dir = ENV["CLAUDE_PLUGIN_DATA"].to_s if data_dir.empty?
        marker_dir = data_dir.empty? ? Dir.tmpdir : data_dir
        FileUtils.mkdir_p(marker_dir) unless Dir.exist?(marker_dir)
        marker = File.join(marker_dir, "okf-plugin-#{session.empty? ? "nosession" : session}.notified")
        return 0 if File.exist?(marker)

        File.write(marker, "")
        emit(stdout, "systemMessage" => "okf plugin: the `okf` CLI is not installed, so bundle curation is off. Run /okf:gem to set it up.")
        0
      end

      # ── plumbing ─────────────────────────────────────────────────────────────

      def relative(file, root)
        file[(root.length + 1)..-1]
      end

      def emit(stdout, payload)
        stdout.puts(JSON.generate(payload))
      end

      def parse(stdin)
        JSON.parse(stdin.read)
      rescue JSON::ParserError
        nil
      end
    end
  end
end

if __FILE__ == $PROGRAM_NAME
  begin
    OKF::PluginHook::Curate.run
  rescue StandardError
    # Silence is the failure mode: a broken hook must not break the session.
  end
  exit 0
end

File: plugins/ai-transmute/hooks/curate.py
#!/usr/bin/env python3
"""Fail-open advisory hook for OKF bundles and AI Transmute jobs."""

from __future__ import annotations

import json
import os
import subprocess
import sys
from pathlib import Path


MAX_LINES = 12


def truthy(name: str) -> bool:
    value = os.environ.get(name, "").strip().lower()
    return bool(value and value not in {"0", "false", "no", "off"})


def edited_path(event: dict) -> Path | None:
    tool_input = event.get("tool_input")
    if not isinstance(tool_input, dict):
        return None
    value = tool_input.get("file_path") or tool_input.get("path")
    return Path(value).expanduser().resolve() if isinstance(value, str) else None


def suppressed(path: Path) -> bool:
    try:
        return path.is_file() and "okf-disable" in path.read_text(encoding="utf-8", errors="ignore")
    except OSError:
        return False


def walk_up(path: Path):
    current = path.parent if path.suffix else path
    while True:
        yield current
        if current.parent == current:
            return
        current = current.parent


def is_okf_root(path: Path) -> bool:
    index = path / "index.md"
    if not index.is_file():
        return False
    try:
        return "okf_version" in index.read_text(encoding="utf-8", errors="ignore")[:512]
    except OSError:
        return False


def run(command: list[str]) -> tuple[int, str]:
    try:
        proc = subprocess.run(command, capture_output=True, text=True, timeout=15, check=False)
        return proc.returncode, (proc.stdout or proc.stderr).strip()
    except (OSError, subprocess.SubprocessError):
        return 0, ""


def context_for(path: Path) -> str | None:
    for parent in walk_up(path):
        if (parent / "job.json").is_file():
            root = os.environ.get("PLUGIN_ROOT") or os.environ.get("CLAUDE_PLUGIN_ROOT")
            if not root:
                return None
            code, output = run([sys.executable, str(Path(root) / "scripts" / "ai_transmute.py"), "validate", str(parent), "--json"])
            if code or '"valid": false' in output.lower():
                return "AI Transmute validation:\n" + "\n".join(output.splitlines()[:MAX_LINES])
            return None
        if is_okf_root(parent):
            messages = []
            for verb in ("validate", "lint"):
                code, output = run(["okf", verb, str(parent), "--json"])
                if code or (output and '"findings":[]' not in output.replace(" ", "")):
                    messages.extend(output.splitlines())
            if messages:
                return "OKF curation advisory:\n" + "\n".join(messages[:MAX_LINES])
            return None
    return None


def main() -> int:
    try:
        if truthy("AI_TRANSMUTE_HOOK_DISABLED"):
            return 0
        event = json.load(sys.stdin)
        path = edited_path(event)
        if path is None or suppressed(path):
            return 0
        context = context_for(path)
        if context:
            print(json.dumps({"hookSpecificOutput": {"hookEventName": "PostToolUse", "additionalContext": context}}))
    except Exception:
        pass
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
